client/comms/wsconn.go

```diff
@@ -64,6 +65,21 @@ func (cs ConnectionStatus) String() string {
 	}
 }
 
+// invalidCertRegexp is a regexp that helps check for non-typed x509 errors
+// caused by or related to an invalid cert.
+var invalidCertRegexp = regexp.MustCompile(".*(unknown authority|not standards compliant|not trusted)")
+
+// isErrorInvalidCert checks if the provided error is one of the different
+// variant of an invalid cert error returned from the x509 package or is
+// ErrInvalidCert.
+func isErrorInvalidCert(err error) bool {
+	var invalidCertErr x509.CertificateInvalidError
+	var unknownCertAuthErr x509.UnknownAuthorityError
+	var hostNameErr x509.HostnameError
+	return errors.Is(err, ErrInvalidCert) || errors.As(err, &invalidCertErr) || errors.As(err, &hostNameErr) ||
```

Comment:
OK to include `ErrInvalidCert` here for completeness I suppose, but this  helpers purpose is to detect when the caller should return `ErrInvalidCert`.  It won't ever get `ErrInvalidCert` itself, or am I missing something?

Reply:
Oh, I had intend to use it in the tests too.

Reply:
I don't see it being useful in the tests.  The test are testing the exported `(*wsConn).Connect` method, ensuring that they return the *exported* error types (`ErrInvalidCert` and `ErrCertRequired`) as expected.

Reply:
The error messages from connect can also be supplemented with the verbose error message while maintaining the wrapped type:

```diff
diff --git a/client/comms/wsconn.go b/client/comms/wsconn.go
index fb1a8123e..43d3d599a 100644
--- a/client/comms/wsconn.go
+++ b/client/comms/wsconn.go
@@ -231,9 +231,9 @@ func (conn *wsConn) connect(ctx context.Context) error {
                if isErrorInvalidCert(err) {
                        conn.setConnectionStatus(InvalidCert)
                        if conn.tlsCfg == nil {
-                               return ErrCertRequired
+                               return dex.NewError(ErrCertRequired, err.Error())
                        }
-                       return ErrInvalidCert
+                       return dex.NewError(ErrInvalidCert, err.Error())
                }
                conn.setConnectionStatus(Disconnected)
                return err
```

Otherwise with this PR's change, you'd never see the "not standards compliant" text even though we would catch it as a cert error and not retry.

Reply:
Yeah, the extra error detail would be useful for maybe future debugging.